### PR TITLE
Fix bug of polygon calculation for voronoiplots (#234)

### LINKED ISSUE #234: [BUG]: Failing to calculate polygons for a voronoi plot
**Describe the bug**

The calculation of polygons for a voronoi plot fails for a specific choice of points (i.e. a grid on a triangle). The issue is inside `grow_polygon_outside_of_box`, where the function returns `Inf`.  I have implemented a workaround locally and will create a PR which I will link to this issue.

**Reproducer**

You can reproduce the issue when trying to do a voronoi plot as below:

```julia
using CairoMakie
using LinearAlgebra

function generatePoints(n)
    points = Vector{NTuple{3,Float64}}(undef, n * (n + 1) ÷ 2)
    m = 1
    for (k, i) in enumerate(range(0, 1, n))
        for j in range(i, 1, n - k + 1)
            px = i
            py = j - i
            pz = 1 - j
            points[m] = (px, py, pz)
            m += 1
        end
    end
    return points
end
function projection(point; origin=[1 / 3, 1 / 3, 1 / 3], e1=normalize(cross(normalize([0.0, 0.0, 1.0] .- [1 / 3, 1 / 3, 1 / 3]), normalize([1 / 3, 1 / 3, 1 / 3]))), e2=normalize([0.0, 0.0, 1.0] .- [1 / 3, 1 / 3, 1 / 3]))
    return sum(e1 .* (point .- origin)), sum(e2 .* (point .- origin))
end

points = generatePoints(3)
points2d = [projection(p) for p in points]
voronoiplot([p[1] for p in points2d], [p[2] for p in points2d], randn(size(points2d)))
```
**Environment information**

```julia
using InteractiveUtils
versioninfo()
```

> Julia Version 1.12.2
> Commit ca9b6662be4 (2025-11-20 16:25 UTC)
> Build Info:
>   Official https://julialang.org release
> Platform Info:
>   OS: macOS (arm64-apple-darwin24.0.0)
>   CPU: 12 × Apple M3 Pro
>   WORD_SIZE: 64
>   LLVM: libLLVM-18.1.7 (ORCJIT, apple-m3)
>   GC: Built with stock GC
> Threads: 1 default, 1 interactive, 1 GC (on 6 virtual cores)
> Environment:
>   JULIA_EDITOR = code

and

```julia
using Pkg
Pkg.status()
```

> Status `/private/var/folders/bb/tp8g2s_95jb3x146dfglhw_h0000gn/T/jl_CI63nn/Project.toml`
>   [13f3f980] CairoMakie v0.15.7

### PR DESCRIPTION
**Pull request by maltepuetz**
This should fix: https://github.com/JuliaGeometry/DelaunayTriangulation.jl/issues/234

I added the internal function `_grow_polygon_outside_of_box`, which is essentially `grow_polygon_outside_of_box`, but takes a type (e.g. `::Float64`) as additional input. 

The original function `grow_polygon_outside_of_box` is now a wrapper for the new internal function. It runs the internal function with `::Float64` (this should have the same behavior as before) and if this fails (i.e. if it returns `new_points`with values that are `Inf`, it runs the internal function again with `t`, `p` and `q` being of type `::BigFloat`.

This ensures higher precision and fixed the issue. Creating the wrapper and running the Float64 version first ensures that performance does not decrease from this fix. 

I almost entirely copied the docstring of the original function for the internal function. Not sure if you want to keep it like that? Also let me know if you have a better idea of fixing this! :)

### CODE REVIEW

File: src/algorithms/voronoi/clipped_coordinates.jl [hunk 1]
```diff
@@ -220,32 +220,78 @@ Truncates the unbounded edges of the `i`th polygon of `vorn` so that the line co
...
 - `new_points`: The new points of the polygon. This is not a circular vector.
 """
-function grow_polygon_outside_of_box(vorn::VoronoiTessellation, i, bounding_box, predicates::AbstractPredicateKernel = AdaptiveKernel())
+function grow_polygon_outside_of_box(vorn::VoronoiTessellation, i, bounding_box, predicates::AbstractPredicateKernel=AdaptiveKernel())
+    # try default Float64 first
+    new_vertices, new_points =
+        _grow_polygon_outside_of_box(Float64, vorn, i, bounding_box, predicates)
+
+    # if the default produced Inf values, rerun with BigFloat
+    has_inf = any(p -> any(isinf, p), new_points)
```
Comment (DanielVandH) [inline src/algorithms/voronoi/clipped_coordinates.jl]: I wonder if it would be better to do `!isfinite` instead of `isinf`? Since this also covers the case where it accidentally produces a `NaN`, i.e.

```diff
- has_inf = any(p -> any(isinf, p), new_points)
+ has_inf = any(p -> any(!isfinite, p), new_points)
```

**Comment by DanielVandH** [comment]
> Also let me know if you have a better idea of fixing this! :)

I had some ideas to try and improve this function but, rather than letting this fix get in the way of me waiting till I might eventually have time to even begin looking into it, this is a great help! I appreciate it.

File: src/algorithms/voronoi/clipped_coordinates.jl (hunk 1 above)
Comment (DanielVandH) [reply src/algorithms/voronoi/clipped_coordinates.jl]: (I would probably prefer something being added here https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/main/src/geometric_primitives/points.jl, like an `allfinite(points)` method that does e.g.

```julia
function allfinite(points)
    for p in each_point(points)
        all(isfinite, getxy(p)) || return false
    end
    return true
end
```

since it's easier for me to keep a unit test for this step, then `has_inf = !allfinite(new_points)`. But that's up to you.)

**Comment by maltepuetz** [comment]
> Can you also add a test for your #234 please?

Will do the test tomorrow, not sure where to put it. Should I put it as a new testset inside https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/main/test/voronoi/voronoi.jl?

**Comment by DanielVandH** [comment]
> Should I put it as a new testset inside https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/main/test/voronoi/voronoi.jl?

Yeah, there's good. And for `allfinite` maybe some tests at the end of

https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/main/test/interfaces/points.jl

would be good. I think these tests are sufficient

```julia
@testset "allfinite" begin
    # All finite points
    @test allfinite([[2.0, 3.5], [1.7, 23.3], [-1.0, 0.0]])
    @test allfinite([(2.0, 3.5), (1.7, 23.3), (-1.0, 0.0)])
    @test allfinite([2.0 1.7 -1.0; 3.5 23.3 0.0])
    @test allfinite(((2.0, 3.5), (1.7, 23.3), (-1.0, 0.0)))
    
    # With Inf
    @test !allfinite([[2.0, 3.5], [Inf, 23.3], [-1.0, 0.0]])
    @test !allfinite([(2.0, 3.5), (Inf, 23.3), (-1.0, 0.0)])
    @test !allfinite([2.0 Inf -1.0; 3.5 23.3 0.0])
    @test !allfinite(((2.0, 3.5), (Inf, 23.3), (-1.0, 0.0)))
    
    # With NaN
    @test !allfinite([[2.0, 3.5], [1.7, NaN], [-1.0, 0.0]])
    @test !allfinite([(2.0, 3.5), (1.7, NaN), (-1.0, 0.0)])
    @test !allfinite([2.0 1.7 NaN; 3.5 23.3 0.0])
    @test !allfinite(((2.0, 3.5), (1.7, NaN), (-1.0, 0.0)))
    
    # With both Inf and NaN
    @test !allfinite([[2.0, 3.5], [Inf, 23.3], [NaN, 0.0]])
    
    # Edge cases
    @test allfinite(Vector{Vector{Float64}}())  # vacuous truth
    @test allfinite(Matrix{Float64}(undef, 2, 0))  # vacuous truth
    @test allfinite([[1.0, 2.0]])
    @test !allfinite([[Inf, 2.0]])
    
    # First point non-finite
    @test !allfinite([[Inf, 1.0], [2.0, 3.0]])
    @test !allfinite([[NaN, 1.0], [2.0, 3.0]])
    
    # Last point non-finite
    @test !allfinite([[1.0, 2.0], [3.0, 4.0], [5.0, Inf]])
    @test !allfinite([[1.0, 2.0], [3.0, 4.0], [NaN, 6.0]])
end
```

**Comment by DanielVandH** [comment]
Your voronoi call isn't the same as the one that voronoiplot uses (it has some clipping it used when it gets the polygon coordinates, that's the difference you're seeing). I'll try and take a look at it myself soon since that might be a bit simpler.

**Comment by DanielVandH** [comment]
> Yes, will do. Should I replace the current tests or just keep them?

Just replace the tests with an appropriate reference test, it's fine. Keep the `allfinite` tests obviously (with `using DelaunayTriangulation: allfinite` at the top of that test file).

**Comment by DanielVandH** [comment]
Thanks!
